### PR TITLE
[Shortest Paths Algorithm] Bidirectional AStar

### PR DESCRIPTION
**Pull request by Shushman**
This PR implements and tests a simple bidirectional A* search algorithm from Goldberg et al "Computing Point-to-Point Shortest Paths from External Memory" (https://archive.siam.org/meetings/alenex05/papers/03agoldberg.pdf)

- The call to shortest paths is exactly the same as for AStar except with an alg::ShortestPaths.BidirAStar, which requires both forward and backward heuristics (defaults to zero). The search returns an AStarResult object with the path and distance.

- For the tests, I have the same tests as for A*. For further correctness, I informally tested the shortest path results on 100 different random queries to a road network graph and confirmed the paths were exactly the same as with A*.

In the next comment, I've pasted a timing script with some results

### CODE REVIEW

**Comment by Shushman** [comment]
I ran the following timing script on my on my 2018 Macbook Pro with Julia 1.5.

```
using Random
using Statistics
using BenchmarkTools
using LinearAlgebra
using Infiltrator
using LightGraphs

const RNG = MersenneTwister(12345)

function main(num_verts::Int64=5000, num_edges::Int64=50000, num_trials=50)

    # Sample 1000 points in [0,1]^2 grid
    vset = rand(RNG, num_verts, 2)
    graph = LightGraphs.SimpleDiGraph(num_verts, num_edges, rng=RNG)
    distmx = reshape( [norm(vset[i, :] - vset[j, :]) for i = 1:num_verts for j = 1:num_verts] , num_verts, num_verts)

    # For each trial, compute paths and times
    astar_times = Float64[]
    bidir_times = Float64[]

    for t = 1:num_trials

        source = rand(RNG, 1:num_verts)
        target = rand(RNG, 1:num_verts)

        # Euclidean heuristic; symmetric
        heuristic(u, t) = norm(vset[u, :] - vset[t, :])
        astar_alg = LightGraphs.ShortestPaths.AStar(heuristic)
        bidir_alg = LightGraphs.ShortestPaths.BidirAStar(heuristic, heuristic)

        astar_time = @elapsed LightGraphs.ShortestPaths.shortest_paths(graph, source, target, distmx, astar_alg)
        bidir_time = @elapsed LightGraphs.ShortestPaths.shortest_paths(graph, source, target, distmx, bidir_alg)


        push!(astar_times, astar_time)
        push!(bidir_times, bidir_time)
    end

    println("AStar Times - Mean: ", mean(astar_times)," ; Median: ", median(astar_times))
    println("Bidir Times - Mean: ", mean(bidir_times)," ; Median: ", median(bidir_times))
end

main()
```

```
julia> include("bmark_bidir_astar.jl")
AStar Times - Mean: 0.0021572885200000004 ; Median: 0.001119882
Bidir Times - Mean: 0.00199417326 ; Median: 0.0017548454999999998
```

**Comment by sbromberger** [comment]
Any way you can get codecov up a bit? You're almost there :)

**Comment by sbromberger** [comment]
Also, let's bikeshed this. Does it make sense to make this a parameter to `AStar()` (like, e.g., `AStar(bidir=true)`)?

**Comment by Shushman** [comment]
Ah I see the tests don't check one specific conditional. I'll come up with a few more to put in there.

On the point of rolling into `AStar`, it could make sense for the current version that is in there, where the `bwd_heuristic` is the only additional parameter (the `fwd_heuristic` is the same as the `heuristic` of AStar). But another knob we could eventually add is the alternation strategy between forward and backward searches. Currently, it is the simplest uninformed one of flipping after every iteration. But the literature has other more informed strategies that look at how the respective searches are proceeding and adjust accordingly. In the long run, we might want that to be an argument to `BidirAStar`, at which point it will look very different. For now, though, I am happy to use `AStar(heuristic, bidirheuristic)` if you prefer.

File: src/ShortestPaths/bidir-astar.jl [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+
+        # Forward or backward search based on bit
+        # Terminate when expanded vertex has been scanned by other search
+        if search_bit == 0 && !isempty(fwd_frontier)
+            (cost_so_far, u) = dequeue!(fwd_frontier)
+
+            if cost_so_far + alg.fwd_heuristic(u, t) > best_path_cost
+                return AStarResult(best_path, best_path_cost)
+            end
```
Comment (sbromberger) [inline src/ShortestPaths/bidir-astar.jl:83]: make this a single line:
```
cost_so_far.... > best_path_cost && return AStarResult(....)
```

File: src/ShortestPaths/bidir-astar.jl [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                end # fwd_colormap[v] < 2
+            end # v in outneighbors
+            fwd_colormap[u] = 2
+
+        elseif search_bit == 1 && !isempty(bwd_frontier)
+            (cost_so_far, u) = dequeue!(bwd_frontier)
+            
+            if cost_so_far + alg.bwd_heuristic(u, s) > best_path_cost
+                return AStarResult(best_path, best_path_cost)
+            end
```
Comment (sbromberger) [inline src/ShortestPaths/bidir-astar.jl:120]: See above for 1-liner

File: src/ShortestPaths/bidir-astar.jl [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    bwd_parents = zeros(U, nvg)
+    bwd_colormap = zeros(UInt8, nvg)
+    bwd_colormap[t] = 1
+
+    # TODO (v2): There should be a more code-efficient way to alternate the searches
+    # where the outneighbors or inneighbors are called and the corresponding bookkeeping is used
+
+    @inbounds while !isempty(fwd_frontier) || !isempty(bwd_frontier)
+
+
```
Comment (sbromberger) [inline src/ShortestPaths/bidir-astar.jl:75]: reduce blank lines here, please

**Review by sbromberger — COMMENTED** [review]
This looks great; thanks! A few minor edits and we should be good to go (don't worry about failing nightly tests).

File: src/ShortestPaths/bidir-astar.jl (hunk 1 above)
Comment (Shushman) [reply src/ShortestPaths/bidir-astar.jl]: Done! Can I ask, is this for efficiency or to match the style in LightGraphs? Or both?

**Comment by Shushman** [comment]
I've addressed the edits you asked for, @sbromberger

**Review by sbromberger — APPROVED** [review]

File: src/ShortestPaths/bidir-astar.jl (hunk 1 above)
Comment (sbromberger) [reply src/ShortestPaths/bidir-astar.jl]: To match the style. It used to be more performant; unsure whether it still is, but have a look at https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/CONTRIBUTING.md, bullet 10.

**Comment by sbromberger** [comment]
Wonderful! Thank you very mucy, @Shushman !

**Pull request merged**